README.md
ZimbraManager
=============
Zimbra commandline administration with Perl and SOAP. 

This tool runs much faster than calling 'zmprov' from the command line.

Use Case
--------
If you like to manage user accounts in scripts you can call the 'zmprov'
command of Zimbra. Unfortunatly always a java environment will be started
which takes multiple seconds for a hugh amount (>10'000) of accounts.

This ZimbraManager is a Mojo Web Service which keeps a session for user
management open and allows to set run SOAP requests based on user REST
requests as input.

Usage
-----

# ZimbraManager::SOAP::Friendly;

    use ZimbraManager::SOAP::Friendly;

    my $authToken = 'VERY_LONG_TOKEN_LINE_FROM_SESSION';
    my $action = 'createAccount';
    my $args = {
        uid                => 'rplessl',
        defaultEmailDomain => 'oetiker.ch',
        givenName          => 'Roman',
        surName            => 'Plessl',
        country            => 'CH',
        displayName        => 'Roman Plessl',
        localeLang         => 'de',
        cosId              => 'ABCD-EFGH-1234',
    };
    my $namedParameters = {
        action    => $action,
        args      => $args,
        authToken => $authToken,
    };
    my ($ret, $err) = $self->soap->callFriendly($namedParameters);


Installation
------------
Build infrastructure to run the framework

    $ cd /opt
    $ git clone https://github.com/rplessl/ZimbraManager
    $ cd ZimbraManager
    $ ./setup/build-perl-modules.sh

Fetch WSDLs and XML Schemas

    $ ./bin/get-wsdl.sh zimbra.example.com

Usage
----- 

    $ ./bin/zimbra-manager.pl prefork

or 

    $ ./bin/zimbra-manager.pl daemon


Deployment
----------
### RHEL

An init.d and sysconfig configuration file is located in setup/rhel

    $ cp setup/rhel/rc.d/init.d/ZimbraManager   /etc/rc.d/init.d/
    $ chmod 755 /etc/rc.d/init.d/ZimbraManager
    $ cp setup/rhel/sysconfig/ZimbraManager     /etc/sysconfig/
    $ cp setup/rhel/logrotate.d/ZimbraManager   /etc/logrotate.d/

    $ chkconfig  --add ZimbraManager
    $ service ZimbraManager restart


LICENSE
--------
GPLv3 license (see LICENSE)

Contact info
------------
Roman Plessl <roman@plessl.info>

http://roman.plessl.info/


CHANGES
2017-02-24 - zaucker - Cleanup log out

2015-01-15 - zaucker - Add getQuotaUsage

2014-06-29 - rplessl - Enhance and fix API as lesson learned by real world 
                       usage

2014-05-08 - rplessl - Changed Interface to named Parameter Interface
                       Added callLegacy() and callFriendlyLegacy() for old API

2014-05-07 - rplessl - added more functions in ZimbraManager::SOAP::Friendly

2014-05-01 - rplessl - fix RHEL init file 
                       fix GetOpt parameter parsing


2014-04-29 - rplessl - Added ZimbraManager::SOAP::Friendly and improve API

2014-03-19 - rplessl - initial Mojo App

2014-03-19 - rplessl - rewritten tool to run faster
                       faster will be done by using and daemon
                       idea: use Mojolicious Perl daemon and 
                       communicate with REST interface to communicator
                        

2014-03-19 - rplessl - initial SOAP tool 
		       proof of concept


test/zimbra-admin-proof-of-concept.pl
#!/usr/bin/env perl

require 5.014;
use strict;

use FindBin;
use lib "$FindBin::Bin/../thirdparty/lib/perl5";

use Data::Dumper;
$Data::Dumper::Indent = 1;

use XML::Compile::WSDL11;
use XML::Compile::SOAP11;
use XML::Compile::Transport::SOAPHTTP;

my $debug = 0;

my $service_port = 'zcsAdminServicePort';
my $service_name = 'zcsAdminService';

my $wsdlFile = "$FindBin::Bin/../etc/ZimbraService.wsdl";
if ($debug) { 
	warn "wsdlFile=$wsdlFile";
}

my $wsdlXml = XML::LibXML->new->parse_file($wsdlFile);

my $zimbraServices;
my $wsdlServices = $wsdlXml->getElementsByTagName( 'wsdl:service' );
for my $service (@$wsdlServices) {
	my $servicename = $service->getAttribute( 'name' );
	my $port        = $service->getElementsByTagName( 'wsdl:port' )->[0];	
	my $address     = $port->getElementsByTagName( 'soap:address' )->[0];
	my $location    = $address->getAttribute( 'location' );
	$zimbraServices->{$servicename} = $location;
}

if ($debug) {
	for my $servicename (keys %$zimbraServices) {
		my $service_address = $zimbraServices->{$servicename};
		   $service_address =~ m/^(https|http):\/\/(.+?):(.+?)\//;
		my $server_protocol = $1;
		my $server_host     = $2;
		my $server_port     = $3;
	
		warn "servicename=",	 $servicename;
		warn "server_protocol=", $server_protocol;
		warn "server_host=",     $server_host;
		warn "server_port=",     $server_port;
		warn "service_address=", $service_address;    	
	}
}

my $wsdl = XML::Compile::WSDL11->new($wsdlXml);
for my $xsd (glob "$FindBin::Bin/../etc/*.xsd") {
	if ($debug) {
		warn "Import", $xsd;
	}
	$wsdl->importDefinitions($xsd);
}

# redirect the endpoint as specified in the WSDL to our own server.
my $transporter = XML::Compile::Transport::SOAPHTTP->new(
	address    => $zimbraServices->{$service_name},
	keep_alive => 1,
);

# enable cookies for zimbra Auth
my $ua = $transporter->userAgent();
$ua->cookie_jar({ file => "$ENV{HOME}/.cookies.txt" });

my $send = $transporter->compileClient( port => $service_port );

# Compile all service methods
my $soapOps = {};
for my $soapOp ( $wsdl->operations( port => $service_port ) ) {
	if ($debug) {
		my $msg = sprintf "got soap operation %s", $soapOp->name;
		warn $msg;
	}
	$soapOps->{ $soapOp->name } =
		$wsdl->compileClient( 	$soapOp->name,
								port      => $service_port,
								transport => $send,
								service   => $service_name );
}

auth();
print Dumper getAccountInfo('user@zimbra.example.com');
print Dumper getAllAccounts();


### call functions ### 
# important ... SOAP functions are starting with a lower capital,
# instead of documented at 
# http://wiki.zimbra.com/wiki/SOAP_API_Reference_Material_Beginning_with_ZCS_8.0
sub auth {
	call($soapOps, 
		 'authRequest', 
		 { persistAuthTokenCookie => 1, 
		   password => 'PASSWORD', 
			account =>  { 
				 by => 'name', 
				  _ => 'admin'}} );
}

sub getAccountInfo {
	my $user = shift;
	my $ret;
	$ret = call($soapOps, 
		 'getAccountInfoRequest', 
		 { account => { 
					by => 'name', 
					_  => $user}});
	return $ret;
}

sub getAllAccounts {
	my $allAccounts = 
	call($soapOps, 
		 'getAllAccountsRequest', 
		 { server => { 
				   by => 'name', 
					_ => 'zimbra.example.com' }, 
		   domain => { 
				   by => 'name', 
					_ => 'zimbra.example.com' }});

	my $accounts;
	for my $za ( @{ $allAccounts->{account} } ) {
		my $name = $za->{name};
		my $id = $za->{id};
		my %kv = map { $_->{'n'} => $_->{'_'} } @{$za->{a}};
		$accounts->{$name} = {
			'name' => $name,
			'id' => $id,
			'kv' => \%kv,
		}
	}
	return $accounts;
}

### generic call function ###	
sub call {
	my ($soapOps, $action, $args) = @_;    
	my %argHash;
	@argHash{ keys %$args }  = values %$args;    

	my ( $response, $trace ) = $soapOps->{$action}->(\%argHash);
	#warn Dumper ("call(): response=", $response);
	#warn Dumper ("call(): trace=", $trace);
	return $response->{parameters};
}

__END__

=head1 NAME

zimbra-admin-proof-of-concept.pl - access administative tools of zimbra with perl

=head1 SYNOPSIS

B<zimbra-admin-interface.pl> [I<options>...]


=head1 DESCRIPTION



=head1 LICENSE

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along with
this program.  If not, see L<http://www.gnu.org/licenses/>.

=head1 AUTHOR

S<Roman Plessl E<lt>roman.plessl@oetiker.chE<gt>>

=head1 HISTORY

 2014-03-19 rp Initial Version

=cut


lib/ZimbraManager.pm
package ZimbraManager;

use Mojo::Base 'Mojolicious';

=pod

=head1 NAME

ZimbraManager - A Mojolicious application to manage Zimbra with SOAP

=head1 SYNOPSIS

    use ZimbraManager;

    # Start commands
    Mojolicious::Commands->start_app('ZimbraManager');

=cut

use Mojo::Util qw(dumper);
use Mojo::JSON qw(decode_json encode_json);

use ZimbraManager::SOAP;
use ZimbraManager::SOAP::Friendly;

use HTTP::CookieJar::LWP;

our $VERSION = "0.17";

=head1 ATTRIBUTES

=head2 soap

The ZimbraManager SOAP object

=cut

has 'soap' => sub {
    my $self = shift;
    return ZimbraManager::SOAP::Friendly->new(
        log => $self->log,
        mode => 'full',
        # soapDebug => '1',           # enables SOAP backend communication debugging
        # soapErrorsToConsumer => '1' # returns SOAP error to consumer
    );
};

=head2 log

Mojo Log object

=cut

has log => sub {
    my $self = shift;
    Mojo::Log->new(
        path  => $ENV{MOJO_LOG_FILE}  // '/dev/stderr',
        level => $ENV{MOJO_LOG_LEVEL} // 'debug',
    );
};

=head1 METHODS

All the methods of L<Mojo::Base> plus:

=head2 private functions

Private functions used in the startup function

=head3 renderOutput

Renders the Output in JSON or readable plain text

=cut

my $renderOutput = sub {
    my ($self, $ctrl, $ret, $err, $plain) = @_;
    my $text = $err ? $err : $ret;
    if ($err) {
        $ctrl->res->code(510);
    }
    if ($plain) {
        if (ref $text eq 'HASH') {
            $text = dumper $text;
        }
        $ctrl->render(text => "<pre>$text</pre>") if ($plain);
    }
    else {
        $ctrl->render(json => $text);
    }
};

=head3 buildAuthRequest

Builds auth call for SOAP

=cut

my $buildAuthRequest = sub {
     my $user = shift;
     my $password = shift;
     return (
     'authRequest',
    { persistAuthTokenCookie => 1,
           password => $password,
            account =>  {
                 by => 'name',
                  _ => $user}}
    );
};

=head3 handleZimbraAuth

Handle the authentication to Zimbra and store the Zimbra authentication
token to the Mojolicious session of the consumer. So the auth token will
be transparently taken to from consumer to the Zimbra end system.

=cut

my $handleZimbraAuth = sub {
    my $self     = shift;
    my $ctrl     = shift;
    my $user     = shift;
    my $password = shift;
    my $ret;
    my $err;
    if ($ctrl->session('ZM_ADMIN_AUTH_TOKEN')) {
        $ret = 'true';
    }
    else {
        my ($action, $args) = $buildAuthRequest->($user,$password);
        ($ret, $err) = $self->soap->call({
            action    => $action,
            args      => $args,
            authToken => undef,
        });
        if (!$err) {
           my $token = $ret->{authToken};
           $ctrl->session('ZM_ADMIN_AUTH_TOKEN' => $token);
        }
        $self->log->debug(dumper("User Session",$ctrl->session));
    }
    $ret = { auth => 'Authentication sucessful!' } if ($ret);
    return ($ret, $err);
};

=head2 startup

Calls Zimbra with the given argument and returns the SOAP response as perl hash.

=cut

sub startup {
    my $self = shift;

    $self->secrets(['bb732c382ded15e58eb02bb0fe0e112e']);
    # session is valid for 1 day
    $self->sessions->default_expiration(1*24*3600);
    $self->sessions->cookie_name('zimbra-manager');

    my $r = $self->routes;

    # Special routing for authentication function for session handling
    $r->post('/auth' => sub {
        my $ctrl        = shift;
        my $perl_args   = decode_json($ctrl->req->body);
        my $user        = $perl_args->{'user'};
        my $password    = $perl_args->{'password'};
        my $plain       = $perl_args->{'plain'};
        my ($ret, $err) = $handleZimbraAuth->($self, $ctrl, $user, $password);
        $renderOutput->($self, $ctrl, $ret, $err, $plain);
    });
    $r->get('/auth' => sub {
        my $ctrl        = shift;
        my $user        = $ctrl->param('user');
        my $password    = $ctrl->param('password');
        my $plain       = $ctrl->param('plain');
        my ($ret, $err) = $handleZimbraAuth->($self, $ctrl, $user, $password);
        $renderOutput->($self, $ctrl, $ret, $err, $plain);
    });

    # Friendly calls to ZimbraManager with key / value parmeters
    $r->post('/friendly/:call' => sub {
        my $ctrl        = shift;
        my $action      = $ctrl->param('call');
        my $plain       = $ctrl->param('plain');
        my $perl_args   = decode_json($ctrl->req->body);
        my ($ret, $err) = $self->soap->callFriendly({
            action    => $action,
            args      => $perl_args,
            authToken => $ctrl->session('ZM_ADMIN_AUTH_TOKEN'),
        });
        $renderOutput->($self, $ctrl, $ret, $err, $plain);
    });
    $r->get('/friendly/:call' => sub {
        my $ctrl        = shift;
        my $action      = $ctrl->param('call');
        my $plain       = $ctrl->param('plain');
        my @param_names = $ctrl->param;
        my $params;
        for my $p (@param_names) {
            $params->{$p} = $ctrl->param($p) unless (($p eq 'action') or ($p eq 'plain'));
        }
        my ($ret, $err) = $self->soap->callFriendly({
            action    => $action,
            args      => $params,
            authToken => $ctrl->session('ZM_ADMIN_AUTH_TOKEN'),
        });
        $renderOutput->($self, $ctrl, $ret, $err, $plain);
    });

    # Handle direct SOAP calls with Zimbra SOAP Datastructure
    $r->post('/:call' => sub {
        my $ctrl        = shift;
        my $action      = $ctrl->param('call');
        my $plain       = $ctrl->param('plain');
        my $perl_args   = decode_json($ctrl->req->body);
        my ($ret, $err) = $self->soap->call({
            action     => $action,
            args       => $perl_args,
            authToken  => $ctrl->session('ZM_ADMIN_AUTH_TOKEN')
        });
        $renderOutput->($self, $ctrl, $ret, $err, $plain);
    });

    return 0;
}

1;

__END__

=head1 COPYRIGHT

Copyright (c) 2014 by Roman Plessl. All rights reserved.
Copyright (c) 2014 by OETIKER+PARTNER AG. 

=head1 LICENSE

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along with
this program.  If not, see L<http://www.gnu.org/licenses/>.

=head1 AUTHOR

S<Roman Plessl E<lt>roman@plessl.infoE<gt>>

=head1 HISTORY

 2014-03-20 rp Initial Version
 2014-04-29 rp New API and added handling of sessions
 2014-05-07 rp Added new API with named parameters

=cut


lib/ZimbraManager/User.pm
package ZimbraManager::User;

use Mojo::Base -base;

=head1 NAME

ZimbraManager::User - User session handling for ZimbraManager

=head1 SYNOPSIS

This User session handling module gets instantiated by L<ZimbraManager> 
and provides session handling functionality for ZimbraManager.

    use ZimbraManager::User;

=head1 DESCRIPTION

Helper class for user session handling.

=cut

use Mojo::Util qw(dumper);
use Mojo::Log;

use HTTP::CookieJar::LWP;

use 5.14.0;

has cookiejar => sub {
    my $self = shift;
    my $cookiejar = $self->controller->session('cookieJarContent');
	if (not $cookiejar) {
		$cookiejar = HTTP::CookieJar::LWP->new();
	}
	return $cookiejar;
};
    
1;

__END__

=head1 LICENSE

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along with
this program.  If not, see L<http://www.gnu.org/licenses/>.

=head1 COPYRIGHT

Copyright (c) 2014 by Roman Plessl. All rights reserved.

=head1 AUTHOR

S<Roman Plessl E<lt>roman.plessl@oetiker.chE<gt>>

=head1 HISTORY

 2014-04-28 rp Initial Version
 
=cut

# Emacs Configuration
#
# Local Variables:
# mode: cperl
# eval: (cperl-set-style "PerlStyle")
# mode: flyspell
# mode: flyspell-prog
# End:
#
# vi: sw=4 et


lib/ZimbraManager/SOAP.pm
package ZimbraManager::SOAP;

use Mojo::Base -base;

=pod

=head1 NAME

ZimbraManager::SOAP - class to manage Zimbra with perl and SOAP

=head1 SYNOPSIS

    use ZimbraManager::SOAP;

    has 'soap' => sub {
        my $self = shift;
        return ZimbraManager::SOAP->new(
            log => $self->log,
            mode => 'full' # or 'admin' or 'user'
            # soapDebug => '1', # enables SOAP backend communication debugging
            # soapErrorsToConsumer => '1' # returns SOAP error to consumer

        );
    };

    my $namedParameters = {
        action    => 'FUNCTIONNAME',
        args      => \%DATASTRUCTUREDPARAMS,
        authToken => $authToken,
    };
    my ($ret, $err) = $self->soap->call($namedParameters);

also

    $self->soap->call(
        action    => 'FUNCTIONNAME',
        args      => \%DATASTRUCTUREDPARAMS,
        authToken => $authToken,
    );

is valid

=head1 DESCRIPTION

Helper class for Zimbra adminstration interface.

=cut

use Mojo::Util qw(dumper);
use Mojo::Log;

use IO::Socket::SSL qw( SSL_VERIFY_NONE SSL_VERIFY_PEER );

use XML::Compile::WSDL11;
use XML::Compile::SOAP11;
use XML::Compile::SOAP::Trace;
use XML::Compile::Transport::SOAPHTTP;

use HTTP::CookieJar::LWP;

our $VERSION = "0.17";

=head1 ATTRIBUTES

=head2 Logging and Debugging Attributes

=head3 log

The mojo log object

=cut

has 'log' => sub {
    return Mojo::Log->new();
};

=head3 soapDebug

Enabled SOAP debug output to $self->log

=cut

has 'soapDebug' => sub {
    my $self = shift;
    return 0;
};

=head3 soapErrorsToConsumer

Returns soapErrors to the consumer in the return messages

=cut

has 'soapErrorsToConsumer' => sub {
    my $self = shift;
    return 0;
};

=head2 mode

The zimbra SOAP interface has three modes:

    admin: adminstration access and commands in admin context
    user:  user accesses and commands in user context
    full:  both admin and user

=cut

has 'mode' => sub {
    my $self = shift;
    return 'full';
};

=head2 wsdlPath

Path to WSDL and XML Schema file(s)

=cut

has 'wsdlPath' => sub {
    my $self = shift;
    return "$FindBin::Bin/../etc/wsdl/";
};

=head2 wsdlFile

Select WSDL file according to mode

=cut

has 'wsdlFile' => sub {
    my $self = shift;
    my $wsdlFile;
    my $mode = $self->mode;
    if   ($mode eq 'admin') {
        $wsdlFile = $self->wsdlPath.'ZimbraAdminService.wsdl';
    }
    elsif ($mode eq 'user') {
        $wsdlFile = $self->wsdlPath.'ZimbraUserService.wsdl';
    }
    elsif ($mode eq 'full') {
        $wsdlFile = $self->wsdlPath.'ZimbraService.wsdl';
    }
    else {
        die "no valid mode ($self->mode) for attribute wsdlFile has been set";
    }
    $self->log->debug("wsdlFile=$wsdlFile");
    return $wsdlFile;
};

=head2 wsdlXml

parsed WSDL as XML

=cut

has 'wsdlXml' => sub {
    my $self = shift;
    my $wsdlXml = XML::LibXML->new->parse_file($self->wsdlFile);
    return $wsdlXml;
};

=head2 wsdl

parsed WSDL as perl object with corresponding included XML Schema
and function stubs.

=cut

has 'wsdl' => sub {
    my $self = shift;
    my $wsdlXml = $self->wsdlXml;
    my $wsdl = XML::Compile::WSDL11->new($wsdlXml);
    for my $xsd (glob $self->wsdlPath."*.xsd") {
        $self->log->debug("XML Schema Import of file: $xsd");
        $wsdl->importDefinitions($xsd);
    }
    return $wsdl;
};

=head2 zcsService

Internal WSDL name for selecting mode

=cut

has 'zcsService' => sub {
    my $self = shift;
    my $mode = $self->mode;
    if    ($mode eq 'admin') {
        return 'zcsAdminService';
    }
    elsif ($mode eq 'user') {
        return 'zcsService';
    }
    elsif ($mode eq 'full') {
        return 'zcsAdminService';
    }
    else {
        die "no valid mode ($self->mode) for attribute zcsService has been set";
    }
};

=head2 wsdlReturnParameterName

Name of the return value / structure

In Zimbra 8.0.7 this will be 'params' as defined here:

    <wsdl:definitions>
     ...
     <wsdl:message name="AdminAbortHsmRequestMessage">
      <wsdl:part name="params" element="zimbraAdmin:AbortHsmRequest"/>
     </wsdl:message>
     ...
    </wsdl:definitions>

In Zimbra 8.0.6 and previous versions this has been 'parameters' as defined here:

    <wsdl:definitions>
     ...
     <wsdl:message name="AdminAbortHsmRequestMessage">
      <wsdl:part name="parameters" element="zimbraAdmin:AbortHsmRequest"/>
     </wsdl:message>
     ...
    </wsdl:definitions>

=cut

has 'wsdlReturnParameterName' => sub {
    return 'params';
};

=head2 service

Processed SOAP service URI on the server

=cut

has 'service' => sub {
    my $self = shift;
    my $wsdlXml = $self->wsdlXml;
    my $zimbraServices;
    my $wsdlServices = $wsdlXml->getElementsByTagName( 'wsdl:service' );
    for my $service (@$wsdlServices) {
        my $name         = $service->getAttribute( 'name' );
        my $port         = $service->getElementsByTagName( 'wsdl:port' )->[0];
        my $port_name    = $port->getAttribute( 'name' );
        my $address      = $port->getElementsByTagName( 'soap:address' )->[0];
        my $uri          = $address->getAttribute( 'location' );
           $uri          =~ m/^(https|http):\/\/(.+?):(.+?)\//;
        my $uri_protocol = $1;
        my $uri_host     = $2;
        my $uri_port     = $3;

        $zimbraServices->{$name} = {
            host           => $uri_host,
            name           => $name,
            port_name      => $port_name,
            uri            => $uri,
            uri_host       => $uri_host,
            uri_port       => $uri_port,
            uri_protocol   => $uri_protocol,
        };
    }
    if ($self->log->level eq 'debug') {
        for my $servicename (keys %$zimbraServices) {
            for my $k (keys %{$zimbraServices->{$servicename}}) {
                $self->log->debug("$k=" . $zimbraServices->{$servicename}->{$k});
            }
        }
    }
    return $zimbraServices->{$self->zcsService};
};

=head2 transporter

Transporter for the SOAP calls - we are using LWP and handle these parameters

    - SSL VERIFY HOSTNAME (see IO::Socket::SSL)
    - SSL VERIFY MODE     (see IO::Socket::SSL)
    - HTTP keep_alive

=cut

has 'transporter' => sub {
    my $self = shift;

    my $uri  = $self->service->{uri};

    my $verifyHostname = $ENV{PERL_LWP_SSL_VERIFY_HOSTNAME} // 1;
    my $verifyMode     = $ENV{PERL_LWP_SSL_VERIFY_MODE}     // SSL_VERIFY_PEER;

    # redirect the endpoint as specified in the WSDL to our own server.
    my $transporter = XML::Compile::Transport::SOAPHTTP->new(
        address    => $uri,
        keep_alive => 1,
        # for SSL handling we need our own LWP Agent
        user_agent => LWP::UserAgent->new(
            ssl_opts => { # default is SSL verification on
                # NOTE: PERL_LWP_SSL_VERIFY_MODE is not an "official" env variable
                verify_hostname => $verifyHostname,
                SSL_verify_mode => $verifyMode,
            },
        ),
    );

    $self->log->debug("LWP_SSL_VERIFY_HOSTNAME = $verifyHostname");
    $self->log->debug("LWP_SSL_VERIFY_MODE     = $verifyMode");

    return $transporter;
};

=head2 soapOps

All usable SOAP operations exported by the the SOAP interface and with the selected
mode.

=cut

has 'soapOps' => sub {
    my $self = shift;
    my $soapOps;

    my $port = $self->service->{port_name};
    my $name = $self->service->{name};
    my $send = $self->transporter->compileClient( port => $port );

    # Compile all service methods
    for my $soapOp ( $self->wsdl->operations( port => $port ) ) {
        $self->log->debug("Got soap operation " . $soapOp->name);
        $soapOps->{ $soapOp->name } =
        $self->wsdl->compileClient(
            $soapOp->name,
            port      => $port,
            service   => $name,
            transport => $send,
        );
    }
    return $soapOps;
};

=head1 METHODS

All the methods of L<Mojo::Base> plus:

=head2 call

Calls Zimbra with the given argument and returns the SOAP response as perl hash.

=cut

# sub callLegacy {
#     my $self      = shift;
#     my $action    = shift;
#     my $args      = shift;
#     my $authToken = shift;
#     my $namedParameters = {
#         action    => $action,
#         args      => $args,
#         authToken => $authToken,
#     };
#     return $self->call($namedParameters);
# }

sub call {
    my $self            = shift;
    my $namedParameters = shift;
    if (ref $namedParameters ne 'HASH') {
        $namedParameters = { @_ };
    }
    my $action          = $namedParameters->{action};
    my $args            = $namedParameters->{args};
    my $authToken       = $namedParameters->{authToken};

    my $uri = $self->service->{uri};

    # for each controller request build authentication token with cookieJar
    # HTTP::CookieJar::LWP is compatible with the original LWP cookie
    # mechanism but lightweight

    my $cookieJar = HTTP::CookieJar::LWP->new();
       $cookieJar->add( $uri, "ZM_ADMIN_AUTH_TOKEN=$authToken" ) if (defined $authToken);
    my $ua = $self->transporter->userAgent();
       $ua->cookie_jar($cookieJar);

    $self->log->debug(dumper({
        _function => 'ZimbraManager::Soap::call',
        action    =>$action,
        args      =>$args,
        authToken =>$authToken
    }));

    my ( $response, $trace ) = $self->soapOps->{$action}->($args);
    if ($self->soapDebug) {
        $self->log->debug(dumper ("call(): response=", $response));
        $self->log->debug(dumper ("call(): trace=",    $trace));
    }
    my $err;
    if ( not defined $response ) {
        $err = 'SOAP ERROR from Zimbra: undefined response';
        if ($self->soapErrorsToConsumer) {
            my $trace    = 'trace:    ' . dumper($trace);
            $err .= "\n\n\n$trace\n";
        }
    }
    elsif ( $response->{Fault} ) {
        $err = 'SOAP ERROR from Zimbra: '. $response->{Fault}->{faultstring};
        if ($self->soapErrorsToConsumer) {
            my $response = 'response: ' . dumper($response);
            my $trace    = 'trace:    ' . dumper($trace);
            $err .= "\n\n\n$response\n\n$trace\n";
        }
    }
    return ($response->{$self->wsdlReturnParameterName}, $err);
}

1;

__END__

=head1 COPYRIGHT

Copyright (c) 2014 by Roman Plessl.
Copyright (c) 2014 by OETIKER+PARTNER AG.
All rights reserved.

=head1 LICENSE

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation, either version 3 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along with
this program.  If not, see L<http://www.gnu.org/licenses/>.

=head1 AUTHOR

S<Roman Plessl E<lt>roman@plessl.infoE<gt>>
S<Fritz Zaucker E<lt>fritz.zaucker@oetiker.chE<gt>>

=head1 HISTORY

 2014-03-20 rp Initial Version
 2014-03-27 rp Improved Error and SSL Handling
 2014-04-29 rp Improved API and Session Handing
 2017-02-24 fz Cleanup log output

=cut

# Emacs Configuration
#
# Local Variables:
# mode: cperl
# eval: (cperl-set-style "PerlStyle")
# mode: flyspell
# mode: flyspell-prog
# End:
#
# vi: sw=4 et
